instrumentation/finagle/http/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finagle/Helpers.java

```diff
@@ -0,0 +1,218 @@
+          Option<Context> ref = Contexts.local().get(ContextRef.INSTANCE.getKey());
+          Callable<Future<?>> call =
+              () -> {
+                try {
+                  return (Future<?>) handle.invoke();
+                } catch (Exception e) {
+                  // don't wrap needlessly
+                  throw e;
+                } catch (Throwable e) {
+                  throw new RuntimeException(e);
+                }
+              };
+          if (ref.isDefined()) {
+            // wrap the call if ContextRef contains a set Context
+            call = ref.get().wrap(call);
+          }
+          try {
+            return call.call();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        },
+        ret);
+  }
+
+  public static <C extends Channel> ChannelInitializer<C> wrapServer(ChannelInitializer<C> inner) {
+    return new ChannelInitializerDelegate<C>(inner) {
+
+      @Override
+      protected void initChannel(C channel) throws Exception {
+        // do all the needful up front, as this may add necessary handlers -- see below
+        super.initChannel(channel);
+
+        // the parent channel is the original http/1.1 channel and has the contexts stored in it;
+        // we assign to this new channel as the old one will not be evaluated in the upgraded h2c
+        // chain
+        Deque<ServerContext> serverContexts =
+            channel.parent().attr(AttributeKeys.SERVER_CONTEXT).get();
+        channel.attr(AttributeKeys.SERVER_CONTEXT).set(serverContexts);
+
+        // todo add way to propagate the protocol version override up to the netty instrumentation;
+        //  why: the netty instrumentation extracts the http protocol version from the HttpRequest
+        //  object which in this case is _always_ http/1.1 due to the use of this adapter codec,
+        //  Http2StreamFrameToHttpObjectCodec
+        ChannelHandlerContext codecCtx =
+            channel.pipeline().context(Http2StreamFrameToHttpObjectCodec.class);
+        if (codecCtx != null) {
+          if (channel.pipeline().get(HttpServerTracingHandler.class) == null) {
+            VirtualField<ChannelHandler, ChannelHandler> virtualField =
+                VirtualField.find(ChannelHandler.class, ChannelHandler.class);
+            HttpServerTracingHandler ourHandler =
+                new HttpServerTracingHandler(
+                    NettyServerSingletons.instrumenter(),
+                    NettyHttpServerResponseBeforeCommitHandler.INSTANCE);
+
+            channel
+                .pipeline()
+                .addAfter(codecCtx.name(), ourHandler.getClass().getName(), ourHandler);
+            // attach this in this way to match up with how netty instrumentation expects things
+            virtualField.set(codecCtx.handler(), ourHandler);
+          }
+        }
+      }
+    };
+  }
+
+  public static <C extends Channel> ChannelInitializer<C> wrapClient(ChannelInitializer<C> inner) {
```

Comment:
ideally this should also be handled by the netty instrumentation?

Reply:
Same answer and question as for server.